### ISSUE TITLE
[API Proposal]: InlineArray to Span conversion without Unsafe.As

### ISSUE THREAD

**Issue by jkotas** [dotnet/runtime#105586]
### Background and motivation

C# allows an implicit conversion of InlineArray to Span. To perform this implicit conversion, Roslyn emits calls of several unsafe methods today (`Unsafe.As`, `MemoryMarshal.CreateSpan`). 

Example: https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwHQCUCuAdgC4CWAtgKbYDCA9uQA6kA2lATgMocBupAxpQDOAbmTIoAFnQBZABQBKZAG9k6dQFFSAcwAWxDWyokh6Sjv2HKx4qYC86YJQBmAQ3wtiYpOvSdGroQAPKQkAHzoQgGE6A7megZGlCYivujIAL7iSADaAJKELKGUAILs7K4AnnIAHAoAushCxOz4/MToWglWNkIqar6hHZRJJN4ZQA===

In hindsight, we have realized that allowing Roslyn to produce unverifiable IL for safe C# is very slippery slope. Instead, we would like safe C# constructs to always produce IL that is verifiable using verification rules written in the same spirit as ECMA-335 verification rules.

The proposed API enables Roslyn to perform the InlineArray to Span conversion without use of unsafe or unverifiable IL.

### API Proposal

```csharp
namespace System.Runtime.CompilerServices;

public class RuntimeHelpers
{
    public static Span<Element> InlineArrayAsSpan<InlineArray, Element>(ref InlineArray inlineArray) where InlineArray: allows ref struct, Element: allows ref struct;
}
```

### API Usage

This API is not meant to be used directly by user code. It is emitted by Roslyn to perform InlineArray to Span conversion.

### Risks

The API is intrinsic with non-trivial expansion logic.

**Comment by jkotas** [comment]
Context: https://github.com/dotnet/runtime/pull/104870#discussion_r1684935295

**Comment by jkotas** [comment]
cc @jaredpar @AlekseyTs

**Comment by hamarb123** [comment]
Even though it's only indended to be used by the compiler, perhaps it's better to name the generic parameters with a `T` as is the usual convention.
```csharp
public static Span<TElement> InlineArrayAsSpan<TInlineArray, TElement>(ref TInlineArray inlineArray) where TInlineArray : allows ref struct, TElement : allows ref struct;
```

**Comment by hez2010** [comment]
This API seems too tricky as we don't have a good way to constraint the inlineArray to be an actual InlineArray. 
In the long term the right move would be allowing constants to be present in type parameters so that we can introduce a generic inline array type `ValueArray<T, int Length>` to serve as a fundamental type for all kinds of purpose.

**Comment by jkotas** [comment]
> This API seems too tricky as we don't have a good way to constraint the inlineArray to be an actual InlineArray.

The implementation of the API would throw if `TInlineArray` is not inline array or if the element type does not match.

**Comment by tannergooding** [comment]
Is there something blocking the verifier from recognizing the pattern and understanding its safe for `InlineArray`?

The API proposed here seems fine to me, just noting that I'd expect the verifier could also understand the existing pattern and make it verifiable.

Roslyn could itself also skip the `Unsafe.As` in the case the backing field of the InlineArray was public, as it could just `MemoryMarshal.CreateSpan(ref buffer.element, 8)` (which should be equally trivial to verify).

**Comment by jkotas** [comment]
> Is there something blocking the verifier from recognizing the pattern and understanding its safe for InlineArray?

It is possible, but the verification rule would not be anything like what exists today. All current verification rules are method local. The current pattern used Roslyn is spread over two methods, so the verification rule would have to talk about multiple method bodies at the same time.

**Comment by timcassell** [comment]
> `TElement: allows ref struct`

I thought `Span<T>` doesn't support ref struct elements due to issues. Is that changing?

**Comment by AlekseyTs** [comment]
Here are the helpers that compiler synthesizes to work with inline arrays. Consider adding corresponding intrinsic APIs:

```
System.Span<TElement> InlineArrayAsSpan<TBuffer, TElement>(ref TBuffer buffer, System.Int32 length)
{
  // Code size       13 (0xd)
  .maxstack  2
  IL_0000:  ldarg.0
  IL_0001:  call       ""ref TElement System.Runtime.CompilerServices.Unsafe.As<TBuffer, TElement>(ref TBuffer)""
  IL_0006:  ldarg.1
  IL_0007:  call       ""System.Span<TElement> System.Runtime.InteropServices.MemoryMarshal.CreateSpan<TElement>(scoped ref TElement, int)""
  IL_000c:  ret
}
```

```
System.ReadOnlySpan<TElement> InlineArrayAsReadOnlySpan<TBuffer, TElement>(in TBuffer buffer, System.Int32 length)
{
  // Code size       18 (0x12)
  .maxstack  2
  IL_0000:  ldarg.0
  IL_0001:  call       ""ref TBuffer System.Runtime.CompilerServices.Unsafe.AsRef<TBuffer>(scoped ref readonly TBuffer)""
  IL_0006:  call       ""ref TElement System.Runtime.CompilerServices.Unsafe.As<TBuffer, TElement>(ref TBuffer)""
  IL_000b:  ldarg.1
  IL_000c:  call       ""System.ReadOnlySpan<TElement> System.Runtime.InteropServices.MemoryMarshal.CreateReadOnlySpan<TElement>(scoped ref readonly TElement, int)""
  IL_0011:  ret
}
```

```
ref TElement InlineArrayElementRef<TBuffer, TElement>(ref TBuffer buffer, System.Int32 index)
{
  // Code size       13 (0xd)
  .maxstack  2
  IL_0000:  ldarg.0
  IL_0001:  call       ""ref TElement System.Runtime.CompilerServices.Unsafe.As<TBuffer, TElement>(ref TBuffer)""
  IL_0006:  ldarg.1
  IL_0007:  call       ""ref TElement System.Runtime.CompilerServices.Unsafe.Add<TElement>(ref TElement, int)""
  IL_000c:  ret
}
```

```
ref readonly TElement InlineArrayElementRefReadOnly<TBuffer, TElement>(in TBuffer buffer, System.Int32 index)
{
  // Code size       18 (0x12)
  .maxstack  2
  IL_0000:  ldarg.0
  IL_0001:  call       ""ref TBuffer System.Runtime.CompilerServices.Unsafe.AsRef<TBuffer>(scoped ref readonly TBuffer)""
  IL_0006:  call       ""ref TElement System.Runtime.CompilerServices.Unsafe.As<TBuffer, TElement>(ref TBuffer)""
  IL_000b:  ldarg.1
  IL_000c:  call       ""ref TElement System.Runtime.CompilerServices.Unsafe.Add<TElement>(ref TElement, int)""
  IL_0011:  ret
}
```

```
ref TElement InlineArrayFirstElementRef<TBuffer, TElement>(ref TBuffer buffer)
{
  // Code size        7 (0x7)
  .maxstack  1
  IL_0000:  ldarg.0
  IL_0001:  call       ""ref TElement System.Runtime.CompilerServices.Unsafe.As<TBuffer, TElement>(ref TBuffer)""
  IL_0006:  ret
}
```

```
ref readonly TElement InlineArrayFirstElementRefReadOnly<TBuffer, TElement>(in TBuffer buffer)
{
  // Code size       12 (0xc)
  .maxstack  1
  IL_0000:  ldarg.0
  IL_0001:  call       ""ref TBuffer System.Runtime.CompilerServices.Unsafe.AsRef<TBuffer>(scoped ref readonly TBuffer)""
  IL_0006:  call       ""ref TElement System.Runtime.CompilerServices.Unsafe.As<TBuffer, TElement>(ref TBuffer)""
  IL_000b:  ret
}
```

**Comment by jkotas** [comment]
@AlekseyTs What do you need `InlineArrayElementRef`/`InlineArrayElementRefReadOnly` and `InlineArrayFirstElementRef`/`InlineArrayFirstElementRefReadOnly` for? Can Roslyn just the Span accessor everywhere and leave the unsafe optimizations to the JIT?

**Comment by hamarb123** [comment]
You need to name it something different as you can't overload by return type or by ref / ref readonly, as I'm sure you're aware :)
Presumably just `InlineArrayAsReadOnlySpan`

**Comment by AlekseyTs** [comment]
```InlineArrayElementRef```/```InlineArrayElementRefReadOnly``` used by compiler to access inline array elements when the index is known to be within bounds. Either the index is compiler controlled (`foreach`, etc.), or user-specified index is a constant.

Same for ```InlineArrayFirstElementRef```/InlineArrayFirstElementRefReadOnly```.

> Can Roslyn use the Span accessor everywhere and leave the unsafe optimizations to the JIT?

Technically it can. But it is not clear what level of optimization JIT is going to provide. Even if it will be able to optimize away span creations, will it be able omit bounds checks on indexing through the span, especially in loop scenarios, etc.?

Is there a problem with exposing specialized helpers?

**Comment by jkotas** [comment]
> will it be able omit bounds checks on indexing through the span, especially in loop scenarios, etc.?

Yes, the JIT has the same optimizations for arrays and spans in general.

> Is there a problem with exposing specialized helpers?

Yes. The goal of this proposal is to enable Roslyn to generate safe verifiable code for InlineArrays. The specialized helpers are unsafe since they do not perform bounds checks.

**Comment by VSadov** [comment]
> This API is not meant to be used directly by user code. 

I wonder why? It seems it could be used by user code.

**Comment by VSadov** [comment]
Also, it could be useful to add `Length`

```cs
public static int InlineArrayLength<TInlineArray>(ref readonly TInlineArray inlineArray) where TInlineArray: allows ref struct;
```

**Comment by jkotas** [comment]
> I wonder why? It seems it could be used by user code.

It could be used by user code just like many similar APIs, but it is not meant to.

**Comment by hamarb123** [comment]
> > This API is not meant to be used directly by user code. 
> 
> I wonder why? It seems it could be used by user code. 

My reading of it is that it means it's not designed for end users to use, like with many other RuntimeHelper apis - I don't think that means you're necessarily not supposed to use it, I think it's just more that it's "not for you".

**Comment by VSadov** [comment]
> Also, it could be useful to add Length

Although, I guess one can do `((Span<T>)arr).Length`

**Comment by jkotas** [comment]
> Also, it could be useful to add Length

The API is expected to be type-safe intrinsics that verifies that the types match, throws exception if they do not, and gets the actual length from the type.

The explicit length argument would be an extra burden. It would not help anything. The implementation would have to verify that the passed in length matches the actual length.

**Comment by jkotas** [comment]
> Although, I guess one can do ((Span<T>)arr).Length

Yep.

**Comment by hamarb123** [comment]
> Also, it could be useful to add `Length`
> 
> ```cs
> public static int InlineArrayLength<TInlineArray>(ref readonly TInlineArray inlineArray) where TInlineArray: allows ref struct;
> ```

Presumably the actual parameter is not necessary (just the generic), as this info should be on the type itself.

**Comment by jkotas** [comment]
> Presumably the actual parameter is not necessary (just the generic), as this info should be on the type itself.

This proposal is for BCL APIs. These APIs are not methods injected into the type by the C# compiler.

**Comment by VSadov** [comment]
I think @hamarb123 just noticed that the argument would be redundant. Which is correct. 
If we do it at all, it could be:

```cs
public static int InlineArrayLength<TInlineArray>() where TInlineArray: allows ref struct;
```

**Comment by hez2010** [comment]
Yes. Making it a generic method can help the runtime implement an intrinsic for this so that it can get expanded to a constant length value at compile time. As well in ILLink/NativeAOT it can be expanded at compile time and redundant branches can be removed.

**Comment by bartonjs** [comment]
[Video](https://www.youtube.com/watch?v=Jwa-CuHL8gY&t=1h29m40s)

Looks good as proposed.

There was a late question of "should we also add where TInlineArray : struct"?  And we'll just accept whatever position @jkotas wants to take.

```csharp
namespace System.Runtime.CompilerServices;

public class RuntimeHelpers
{
    public static Span<TElement> InlineArrayAsSpan<TInlineArray, TElement>(ref TInlineArray inlineArray) where TInlineArray: allows ref struct;
    public static ReadOnlySpan<TElement> InlineArrayAsReadOnlySpan<TInlineArray, TElement>(ref readonly TInlineArray inlineArray) where TInlineArray: allows ref struct;

}
```

**Comment by jkotas** [comment]
> should we also add where TInlineArray : struct

Sounds good to me

**Comment by AlekseyTs** [comment]
Have you seen https://github.com/dotnet/runtime/issues/105586#issuecomment-2359585731? Compiler actually uses more helpers, some of them return a reference to a specific element and skipping a span creation. They are used when compiler can perform a boundary check at compile time. For example, when a constant is used as an index, or during a `foreach` loop.

**Comment by hamarb123** [comment]
@AlekseyTs I believe this was answered previously in https://github.com/dotnet/runtime/issues/105586#issuecomment-2362500536 - when the applicable runtime helpers API is available, you should just call that and then call the span indexer on the result of that instead of having the other 2 APIs.

E.g., when the APIs are available, rather than emitting `ref InlineArrayElementRef(ref buffer, i)` with the helper method, you would emit `ref RuntimeHelpers.InlineArrayAsSpan(ref buffer)[i]` for example.

**Comment by jkotas** [comment]
@AlekseyTs If you would like to see a helper that is a shortcut for `ref RuntimeHelpers.InlineArrayAsSpan(ref buffer)[i]` (+ readonly version), we should be able to add it. It would be a small IL size optimization. It would behave the same way as the longer form at runtime.

**Comment by AlekseyTs** [comment]
@jkotas I guess that wouldn't be equivalent to the helpers that compiler uses today. For example, compiler helpers do not perform a boundary check. If we simply switch to span indexing, is JIT going to elide the check in those scenarios? If not, then it is quite possible that the code will become slower.

**Comment by AlekseyTs** [comment]
According to https://github.com/dotnet/runtime/issues/105586#issuecomment-2362500536, it looks like we can assume that JIT will elide the check. Given this, I don't think compiler needs a special helper for indexing.

**Comment by tannergooding** [comment]
It should elide it, because it specifically understands `new Span(ref element. cns)` as giving a length of `cns` so any `new Span(ref element, cns)[cns2]` will elide if `cns2 >= 0 && cns2 < cns`

We might need to consider T0 vs T1 tuning on that, but that's a JIT impl detail for us to think about and shouldn't impact Roslyn.
